onnxruntime/core/providers/dml/DmlExecutionProvider/src/External/DirectMLHelpers/ApiTraits.h

```diff
@@ -2717,4 +2790,658 @@ inline gsl::czstring ToString(DML_OPERATOR_TYPE value)
+    switch (value)
+    {
+    case DML_IS_INFINITY_MODE_EITHER: return "DML_IS_INFINITY_MODE_EITHER";
+    case DML_IS_INFINITY_MODE_POSITIVE: return "DML_IS_INFINITY_MODE_POSITIVE";
+    case DML_IS_INFINITY_MODE_NEGATIVE: return "DML_IS_INFINITY_MODE_NEGATIVE";
+    default:
+        assert(false);
+        return "<unknown>";
+    }
+}
+
+template <>
+inline gsl::czstring ToString(DML_DEPTH_SPACE_ORDER value)
+{
+    switch (value)
+    {
+    case DML_DEPTH_SPACE_ORDER_DEPTH_COLUMN_ROW: return "DML_DEPTH_SPACE_ORDER_DEPTH_COLUMN_ROW";
+    case DML_DEPTH_SPACE_ORDER_COLUMN_ROW_DEPTH: return "DML_DEPTH_SPACE_ORDER_COLUMN_ROW_DEPTH";
+    default:
+        assert(false);
+        return "<unknown>";
+    }
+}
+
+template <>
+inline gsl::czstring ToString(DML_AXIS_DIRECTION value)
+{
+    switch (value)
+    {
+    case DML_AXIS_DIRECTION_INCREASING: return "DML_AXIS_DIRECTION_INCREASING";
+    case DML_AXIS_DIRECTION_DECREASING: return "DML_AXIS_DIRECTION_DECREASING";
+    default:
+        assert(false);
+        return "<unknown>";
+    }
+}
+
+template <>
+inline gsl::czstring ToString(DML_ROUNDING_MODE value)
+{
+    switch (value)
+    {
+    case DML_ROUNDING_MODE_HALVES_TO_NEAREST_EVEN: return "DML_ROUNDING_MODE_HALVES_TO_NEAREST_EVEN";
+    case DML_ROUNDING_MODE_TOWARD_ZERO: return "DML_ROUNDING_MODE_TOWARD_ZERO";
+    case DML_ROUNDING_MODE_TOWARD_INFINITY: return "DML_ROUNDING_MODE_TOWARD_INFINITY";
+    default:
+        assert(false);
+        return "<unknown>";
+    }
+}
+
+template <>
+inline gsl::czstring ToString(DML_RANDOM_GENERATOR_TYPE value)
+{
+    switch (value)
+    {
+    case DML_RANDOM_GENERATOR_TYPE_PHILOX_4X32_10: return "DML_RANDOM_GENERATOR_TYPE_PHILOX_4X32_10";
+    default:
+        assert(false);
+        return "<unknown>";
+    }
+}
+
+template <>
+inline gsl::czstring ToString(DML_MULTIHEAD_ATTENTION_MASK_TYPE value)
+{
+    switch (value)
+    {
+    case DML_MULTIHEAD_ATTENTION_MASK_TYPE_NONE: return "DML_MULTIHEAD_ATTENTION_MASK_TYPE_NONE";
+    case DML_MULTIHEAD_ATTENTION_MASK_TYPE_KEY_SEQUENCE_LENGTH: return "DML_MULTIHEAD_ATTENTION_MASK_TYPE_KEY_SEQUENCE_LENGTH";
+    case DML_MULTIHEAD_ATTENTION_MASK_TYPE_KEY_SEQUENCE_END_START: return "DML_MULTIHEAD_ATTENTION_MASK_TYPE_KEY_SEQUENCE_END_START";
+    case DML_MULTIHEAD_ATTENTION_MASK_TYPE_KEY_QUERY_SEQUENCE_LENGTH_START_END: return "DML_MULTIHEAD_ATTENTION_MASK_TYPE_KEY_QUERY_SEQUENCE_LENGTH_START_END";
+    case DML_MULTIHEAD_ATTENTION_MASK_TYPE_BOOLEAN: return "DML_MULTIHEAD_ATTENTION_MASK_TYPE_BOOLEAN";
+    default:
+        assert(false);
+        return "<unknown>";
+    }
+}
+
+
+template <typename T>
+inline T FromString(std::string_view value)
+{
+    // Clang will instantiate this template even if it isn't used,
+    // so this static_assert will always fire and break the build.
+    static_assert(false, "Not implemented for this type");
+}
+
+inline std::unordered_map<std::string_view, DML_TENSOR_DATA_TYPE> DML_TENSOR_DATA_TYPE_MAP = {
```

Comment:
>inline std::unordered_map [](http://example.com/codeflow?start=0&length=25)

We shouldn't initialize complex data types like `std::map` (which does lots of allocations) directly inline in the header file. Otherwise on DLL load, we'll incur several unnecessary hits.

Also every cpp file which includes this header will absorb the map which needs to be deduplicated by the linker COMDAT.

Using a constexpr type puts it in the readonly section of the PE, like what we do in DML:

```
constexpr StringUtil::WideNameAndIndex mapping[] =
    {
        {L"ARGMAX", DML_REDUCE_FUNCTION_ARGMAX},
        {L"ARGMIN", DML_REDUCE_FUNCTION_ARGMIN},
        {L"AVERAGE", DML_REDUCE_FUNCTION_AVERAGE},
        {L"L1", DML_REDUCE_FUNCTION_L1},
        {L"L2", DML_REDUCE_FUNCTION_L2},
        {L"LOG_SUM", DML_REDUCE_FUNCTION_LOG_SUM},
        {L"LOG_SUM_EXP", DML_REDUCE_FUNCTION_LOG_SUM_EXP},
        {L"MAX", DML_REDUCE_FUNCTION_MAX},
        {L"MIN", DML_REDUCE_FUNCTION_MIN},
        {L"MULTIPLY", DML_REDUCE_FUNCTION_MULTIPLY},
        {L"SUM", DML_REDUCE_FUNCTION_SUM},
        {L"SUM_SQUARE", DML_REDUCE_FUNCTION_SUM_SQUARE},
    };
```

Do we need a map, given name lookup probably isn't the biggest perf impact here, vs a linear scan? If a map is warranted, is there a good place to just initialize them once, or initialize lazily? Maybe we can initialize them inside the corresponding `FromString` function? Maybe we don't even need them?

Reply:
Note compile-time variants of std::map *do* exist, but they are external libraries which would pull in a dependency:
[https://github.com/serge-sans-paille/frozen](https://github.com/serge-sans-paille/frozen)